### ISSUE TITLE
[TRI-1008] Add `io.waitUntil` function to allow a job run to wait for a dynamic condition

### ISSUE THREAD

**Issue by ericallam** [triggerdotdev/trigger.dev#315]
This kind of thing would be really useful in jobs:

```ts
await io.waitUntil("has used invite", {
  condition: async () => {
    const user = await prisma.user.findUniqueOrThrow({
      where: {
        id: payload.userId,
      },
    });

    return user.invitationCodeId !== null;
  },
});
```

This would basically retry the condition function in an interval, and only continue once the condition function returned `true`.

The strategy for how frequent to call the condition function could be configurable:

- Set an interval frequency (minimum would be 60 seconds, max would be 1 day). Would also need some kind of timeout thing like "exit if the condition doesn't return true within 30 days"
- Set an exponential backoff strategy (similar to our task retries) with a maximum of 25 retries

This can be achieved using only subtasks and `io.wait`, something like:

```ts
type WaitUntilOptions = {
  condition: () => Promise<boolean>;
  checkInterval?: number;
};

function waitUntil(
  key: string,
  io: IO,
  options: WaitUntilOptions
): Promise<void> {
  return io.runTask(
    key,
    {
      name: "Wait until",
    },
    async (task) => {
      const result = await options.condition();
      if (result) return;

      //if more than 14 days has passed then throw
      const startedAt = task.startedAt;
      if (!startedAt) return;
      const now = new Date();
      const diff = now.getTime() - startedAt.getTime();
      const days = diff / (1000 * 60 * 60 * 24);
      if (days > 14) {
        throw new Error(
          `Waited for 14 days for condition to be true but it never was.`
        );
      }

      await io.wait(
        `${key}:${new Date().toISOString()}`,
        options.checkInterval ? Math.max(options.checkInterval, 30) : 60 * 60
      );
    }
  );
}
```

But i'd prefer this not to produce a bunch of tasks that are only there for the interval. 

<sub>[TRI-1008](https://linear.app/triggerdotdev/issue/TRI-1008/add-iowaituntil-function-to-allow-a-job-run-to-wait-for-a-dynamic)</sub>

**Comment by zubairov** [comment]
That's a good idea, however I would love to also see that a Job can wait for a specific `event` with possibly a filter on the event payload. A use-case in-mind would be from the marketing / lead generation area where a potential lead want to download a resource / e-book from a provider however before doing that an e-mail double-opt-in should happen, so the Job that is sending an e-book to a newly registered lead would need to wait for opt-in event of that user.

**Comment by ericallam** [comment]
@zubairov yup that's a great idea! And something we definitely want to support (and everything is already in place for this to be supported). Here's my idea for a syntax:

```ts
const event = await io.waitForNextEvent("✋", { name: "opt-in", filter: { userId: [payload.userId] })
```

**Comment by zubairov** [comment]
awesome, one small note - if possible to have a filter as a ``async() =>`` function would be great - this way in the use-case above a job will wait for an opt-in of a _particular_ lead not just _any_ lead

**Comment by zubairov** [comment]
ah, right sorry missed that one, it's an actual value that is controlled for not only the presence of particular field in the event. Got it. That would do. Just thought a function might be more flexible in the long run

**Comment by matt-aitken** [comment]
v2 is deprecated as of Jan 31 2025
